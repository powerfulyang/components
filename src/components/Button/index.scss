$base-color: #90caf9;

.py-ripple-click {
  position: absolute;
  overflow: hidden;
  inset: 0;
  border-radius: inherit;

  .py-ripple {
    position: relative;
    width: 100%;
    aspect-ratio: 1;
    background-color: rgb(0 0 0 / 20%);
    border-radius: 50%;
  }
}
