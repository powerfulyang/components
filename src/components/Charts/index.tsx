import type { FC, PropsWithChildren } from 'react';
import React, { useId, useMemo } from 'react';
import { atomWithImmer } from 'jotai-immer';
import { ChartContext } from '@/components/Charts/ChartContext';

export * from './cartesian';
export { ChartContext };

type ChartProps<T> = {
  width: number;
  height: number;
  viewBox?: {
    x: number;
    y: number;
    width: number;
    height: number;
  };
  data: T[];
  padding?: {
    top?: number;
    right?: number;
    bottom?: number;
    left?: number;
  };
};

type ClipPathProps = {
  id: string;
  x: number;
  y: number;
  width: number;
  height: number;
};
const ClipPath: FC<ClipPathProps> = ({ id, x, y, width, height }) => {
  return (
    <defs>
      <clipPath id={id}>
        <rect x={x} y={y} height={height} width={width} />
      </clipPath>
    </defs>
  );
};

export const Chart = <T extends unknown>({
  width,
  height,
  viewBox,
  data,
  children,
  padding,
}: PropsWithChildren<ChartProps<T>>) => {
  const svgView = viewBox || { width, height, x: 0, y: 0 };
  const id = useId();
  const clipPathId = `clip-path-${id}`;
  const atomInstance = useMemo(() => {
    return atomWithImmer({
      xAxis: {
        size: 0,
        dataKey: '',
      },
      yAxis: {
        size: 0,
        dataKeys: [],
      },
    });
  }, []);
  const ctx = useMemo(() => {
    return {
      atomInstance,
      width,
      height,
      padding: {
        top: padding?.top || 0,
        right: padding?.right || 0,
        bottom: padding?.bottom || 0,
        left: padding?.left || 0,
      },
      data,
    };
  }, [atomInstance, data, height, padding, width]);
  const clipPath = useMemo(() => {
    return {
      id: clipPathId,
      x: ctx.padding.left,
      y: ctx.padding.top,
      width: ctx.width - ctx.padding.left - ctx.padding.right,
      height: ctx.height - ctx.padding.top - ctx.padding.bottom,
    };
  }, [
    clipPathId,
    ctx.height,
    ctx.padding.bottom,
    ctx.padding.left,
    ctx.padding.right,
    ctx.padding.top,
    ctx.width,
  ]);
  return (
    <ChartContext.Provider value={ctx}>
      <svg
        id={id}
        width={width}
        height={height}
        viewBox={`${svgView.x} ${svgView.y} ${svgView.width} ${svgView.height}`}
      >
        <ClipPath {...clipPath} />
        {children}
      </svg>
    </ChartContext.Provider>
  );
};
