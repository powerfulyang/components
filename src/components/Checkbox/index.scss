.py-checkbox {
  position: relative;
  display: inline-block;
  width: 20px;
  height: 20px;
  transition: background 150ms ease;
  border: 2px solid #211076;
  border-radius: 3px;

  &:hover {
    background: rgba(33, 16, 118, 0.5);
  }

  &:after {
    position: absolute;
    top: 1px;
    left: 5px;
    width: 6px;
    height: 12px;
    content: '';
    transition: all 325ms ease;
    transform: rotate(45deg) scale(0);
    opacity: 0;
    border-right: 2px solid #bf6bfe;
    border-bottom: 2px solid #bf6bfe;
  }
}
