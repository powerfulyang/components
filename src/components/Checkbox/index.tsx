import type { FC, PropsWithChildren } from 'react';
import React, { forwardRef, useId } from 'react';
import './index.scss';
import classNames from 'classnames';
import { useIsomorphicLayoutEffect } from '@powerfulyang/hooks';
import { Icon } from '../Icon';

export type CheckboxProps = React.DetailedHTMLProps<
  React.InputHTMLAttributes<HTMLInputElement>,
  HTMLInputElement
> & {
  indeterminate?: boolean;
};

export const Checkbox: FC<PropsWithChildren<CheckboxProps>> = forwardRef<
  HTMLInputElement,
  CheckboxProps
>(({ id, className, children, indeterminate, ...props }, ref) => {
  const elementId = useId();

  const inputId = id || elementId;

  useIsomorphicLayoutEffect(() => {
    if (indeterminate !== undefined) {
      (document.getElementById(inputId) as HTMLInputElement).indeterminate = indeterminate;
    }
  }, [indeterminate, inputId]);

  return (
    <label htmlFor={inputId} className="relative cursor-pointer">
      <input
        {...props}
        ref={ref}
        type="checkbox"
        id={inputId}
        className={classNames('py-checkbox', className)}
      />
      <span className="py-checkbox__marker">
        <Icon className="py-checkbox__checked" type="icon-checked" />
        <Icon className="py-checkbox__indeterminate" type="icon-indeterminate" />
      </span>
      <span className="ml-4 mr-3">{children}</span>
    </label>
  );
});

Checkbox.displayName = 'Checkbox';
