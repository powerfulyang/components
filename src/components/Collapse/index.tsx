import React from 'react';
import AnimateHeight, { AnimateHeightProps } from 'react-animate-height';
import { __dev__ } from '@powerfulyang/utils';
import { forwardRef } from '@/utils/forwardRef';

export interface CollapseOptions {
  /**
   * If `true`, the content will be visible
   */
  isOpen?: boolean;
  /**
   * If `true`, the opacity of the content will be animated
   */
  animateOpacity?: boolean;
  /**
   * The duration of the animation in `ms`
   */
  duration?: number;
  /**
   * The height you want the content in it's collapsed state. Set to `0` by default
   */
  startingHeight?: number | string;
  /**
   * The height you want the content in it's expanded state. Set to `auto` by default
   */
  endingHeight?: number | string;
  /**
   * The function to be called when the collapse animation starts. It provides the `newHeight` as an argument
   */
  onAnimationEnd?(props: { newHeight: number }): void;
  /**
   * The function to be called when the collapse animation ends. It provides the `newHeight` as an argument
   */
  onAnimationStart?(props: { newHeight: number }): void;
}

export interface CollapseProps extends AnimateHeightProps, CollapseOptions {}

export const Collapse = forwardRef<CollapseProps, 'div'>((props, ref) => {
  const {
    isOpen,
    animateOpacity = true,
    onAnimationStart,
    onAnimationEnd,
    duration,
    easing,
    startingHeight = 0,
    endingHeight = 'auto',
    as: Component = 'div',
    ...rest
  } = props;
  return (
    <AnimateHeight
      duration={duration}
      easing={easing}
      animateOpacity={animateOpacity}
      height={isOpen ? endingHeight : startingHeight}
      applyInlineTransitions={false}
      onAnimationStart={onAnimationStart}
      onAnimationEnd={onAnimationEnd}
    >
      <Component ref={ref} {...rest} />
    </AnimateHeight>
  );
});

if (__dev__) {
  Collapse.displayName = 'Collapse';
}
