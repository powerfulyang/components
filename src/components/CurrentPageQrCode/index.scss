.shareQrCode {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 115px;
  height: 115px;
  canvas {
    width: 110px;
    height: 110px;
  }
}
