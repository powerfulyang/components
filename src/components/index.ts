export * from './Collapse';
export * from './Icon';
export * from './Icon/IconTag';
export * from './Icon/ShowQrCode';
export * from './Tooltip';
export * from './CurrentPageQrCode';
export * from './Skeleton';
export * from './TagCloud';
