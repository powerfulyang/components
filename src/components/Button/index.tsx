import type { MouseEvent, ReactElement } from 'react';
import React, { cloneElement, forwardRef, useCallback } from 'react';
import RippleEffect from '@/components/RippleEffect';
import { Icon } from '@/components/Icon';
import { ClassNames, css, keyframes } from '@emotion/react';
import classNames from 'classnames';
import { useTheme } from '@/context/theme';

export type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  ripple?: boolean;
  appearance?: 'primary' | 'secondary' | 'error' | 'default';
  ghost?: boolean;
  rounded?: boolean;
  icon?: ReactElement;
  loading?: boolean;
};

const spin = keyframes`
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
`;

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      children,
      className,
      onClick,
      ripple = true,
      appearance = 'default',
      rounded = false,
      icon,
      loading = false,
      ghost = false,
      ...props
    },
    ref,
  ) => {
    const { disabled = loading } = props;
    const canClick = !disabled && !loading;

    const handleClick = useCallback(
      (e: MouseEvent<HTMLButtonElement>) => {
        if (canClick && onClick) {
          onClick(e);
        }
      },
      [canClick, onClick],
    );

    const theme = useTheme();

    return (
      <button
        type="button"
        {...props}
        ref={ref}
        css={[
          css({
            display: 'inline-flex',
            justifySelf: 'center',
            alignItems: 'center',
            position: 'relative',
            overflow: 'hidden',
            padding: '.65rem 1rem',
            color: 'white',
            cursor: loading ? 'not-allowed' : 'pointer',
            border: 'none',
            borderRadius: rounded ? '9999px' : '0.5rem',
            backgroundColor: theme.colors.transparent,
          }),
          css`
            &:disabled,
            &[data-loading='true'] {
              cursor: not-allowed;
              opacity: 0.5;
              background-color: ${theme.colors.disabled};
            }
            &:not(:disabled) {
              background-color: ${theme.colors[appearance]};
            }
            &:not(:disabled):not([data-loading='true']):hover {
              filter: brightness(1.1);
            }
          `,
          ghost &&
            css`
              &:not(:disabled) {
                border: 1px solid ${theme.colors[appearance]};
                background-color: ${theme.colors.transparent};
              }
            `,
        ]}
        className={className}
        onClick={handleClick}
        data-loading={loading}
      >
        <ClassNames>
          {({ css: CSS }) => {
            return (
              <>
                {loading && (
                  <Icon
                    type="icon-loading"
                    css={CSS`
                      margin-right: 0.6rem;
                      transform-origin: center;
                      animation: ${spin} 1s linear infinite;
                    `}
                  />
                )}
                {!loading &&
                  icon &&
                  cloneElement(icon, {
                    ...icon.props,
                    className: classNames(
                      CSS`
                      margin-right: 0.6rem;

                    `,
                      icon.props.className,
                    ),
                  })}
              </>
            );
          }}
        </ClassNames>
        {children}
        {ripple && canClick && <RippleEffect />}
      </button>
    );
  },
);

Button.displayName = 'Button';
