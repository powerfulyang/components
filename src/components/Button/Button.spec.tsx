import { render } from '@testing-library/react';
import { Button } from '@/components';
import { describe, expect, it } from '@jest/globals';

describe('Button', () => {
  it('should invoke onClick', () => {
    const onClick = jest.fn();
    const wrapper = render(<Button onClick={onClick}>Button</Button>);
    wrapper.getByText('Button').click();
    expect(onClick).toHaveBeenCalled();
  });

  it('should disable click', () => {
    const onClick = jest.fn();
    const wrapper = render(
      <Button onClick={onClick} disabled>
        Button
      </Button>,
    );
    wrapper.getByText('Button').click();
    expect(onClick).not.toHaveBeenCalled();
  });

  it('snapshot', () => {
    const { container } = render(<Button>Button</Button>);
    expect(container).toMatchSnapshot();
  });
});
