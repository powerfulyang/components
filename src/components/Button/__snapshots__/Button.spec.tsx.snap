// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshot 1`] = `
@keyframes animation-0 {
  0% {
    -webkit-transform: scale(0);
    -moz-transform: scale(0);
    -ms-transform: scale(0);
    transform: scale(0);
    opacity: 1;
  }

  50% {
    opacity: 1;
  }

  100% {
    opacity: 0;
    -webkit-transform: scale(2);
    -moz-transform: scale(2);
    -ms-transform: scale(2);
    transform: scale(2);
  }
}

.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  justify-self: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  overflow: hidden;
  padding: .65rem 1rem;
  color: white;
  cursor: pointer;
  border: none;
  border-radius: 0.5rem;
  background-color: transparent;
}

.emotion-0:disabled,
.emotion-0[data-loading='true'] {
  cursor: not-allowed;
  opacity: 0.5;
  background-color: #9ca3af;
}

.emotion-0:not(:disabled) {
  background-color: transparent;
}

.emotion-0:not(:disabled):not([data-loading='true']):hover {
  -webkit-filter: brightness(1.1);
  filter: brightness(1.1);
}

.emotion-1 {
  position: absolute;
  inset: 0;
}

.emotion-1>span {
  position: absolute;
  display: block;
  background-color: rgb(0 0 0 / 8%);
  border-radius: 50%;
  -webkit-transform: scale(0);
  -moz-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-animation: animation-0 700ms;
  animation: animation-0 700ms;
}

<div>
  <button
    class="emotion-0"
    data-loading="false"
    data-projection-id="2"
    type="button"
  >
    Button
    <div
      class="emotion-1"
      role="none"
    />
  </button>
</div>
`;
